### PR TITLE
[MSFT] Add a physical location attribute

### PR DESCRIPTION
**Pull request by teqdruid**
This attribute is intended to be used as an annotation to specify a placement constraint.

### CODE REVIEW

**Comment by teqdruid** [comment]
Is there any way to get an attribute built in ODS (via extending StructAttr) to get printed the dialect printer? Looking at how it works, I don't think so but I may be wrong. This is probably not the intent of this particular attribute creation method.

**Comment by River707** [comment]
Sorry missed this (hectic weekend). The `StructAttr` wrapper around DictionaryAttr isn't intended to be very smart, and is mostly just for building a type-safe wrapper around DictionaryAttr. If you want the benefits of a custom attribute (like parsing/printing, you'll need to define one in your dialect). You can now (mostly, likely a thing or two missing) define Attributes in ODS so the barrier to entry there is lower, e.g.:
  https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinAttributes.td
  https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td

**Comment by teqdruid** [comment]
I didn't realize ODS support for Attributes was added... I couldn't find any mention of it in the ODS documentation. Serves me right for not checking `OpBase.td` before I started!

Thanks for the response!

**Comment by River707** [comment]
No worries, it is fairly recent (like last week/week and a half). The docs for attribute/type building need a restructure in general (within the next week or so) so kind of pushed updating everything until then.

**Comment by teqdruid** [comment]
@River707 I re-coded it using the new ODS support. Nice! Thanks for the info.

File: include/circt/Dialect/MSFT/MSFT.td [hunk 1]
```diff
@@ -33,4 +33,29 @@ def MSFTDialect : Dialect {
...
+  let summary = "Descibes a physical location on a device";
+  let description = [{
+    Annotate a particular entity within an op with the location of the device
+    on an FPGA to which it should mapped.
+  }];
+  let mnemonic = "physloc";
+  let parameters = (ins
+    "DeviceTypeAttr":$devType,
+    "uint64_t":$x, "uint64_t":$y, "uint64_t":$num,
+    "::llvm::StringRef":$entity);
```
Comment (River707) [inline include/circt/Dialect/MSFT/MSFT.td:58]: Drive by, but what is the lifetime of `entity` here? Should this be using `StringRefParameter` instead?

Comment (teqdruid) [reply include/circt/Dialect/MSFT/MSFT.td:58]: D'oh! I was making an assumption which isn't valid. Good catch -- thanks!

Comment (teqdruid) [reply include/circt/Dialect/MSFT/MSFT.td]: Ended up going a different way with this which obviated the need for this parameter.

**Pull request merged**